docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,211 @@
+    timeout: "5m"
+  - type:    "Ready"
+    status:  "False"
+    timeout: "5m"
+  maxUnhealthy: "40%"
+status:
+  currentHealthy: 5
+  expectedMachines: 5
+```
+
+#### MachineHealthCheck controller:
+Watch:
+- Watch machineHealthCheck resources
+- Watch machines and nodes with an event handler e.g controller runtime `EnqueueRequestsFromMapFunc` which returns machineHealthCheck resources.
+
+Reconcile:
+- Fetch all the machines targeted by the MachineHealthCheck and operate over machine/node targets. E.g:
+```
+type target struct {
+  Machine capi.Machine
+  Node    *corev1.Node
+  MHC     capi.MachineHealthCheck
+}
+```
+
+- Calculate the number of unhealthy targets.
+- Compare current number against `maxUnhealthy` threshold and temporary short circuits remediation if the threshold is met.
+- Trigger remediation for unhealthy targets i.e request machines for deletion.
+
+Out of band:
+- The owning controller e.g machineSet controller reconciles to meet number of replicas and start the process to bring up a new machine/node.
+- The machine controller drains the unhealthy node.
+- The machine controller provider deletes the instance.
+- The machine controller deletes the machine.
+
+![Machine health check](./images/machine-health-check/mhc.svg)
+
+### Risks and Mitigations
+
+## Alternatives
+
+Considered to bake this functionality into machineSets. This was discarded as different controllers than a machineSet could be owning the targeted machines. For those cases as a user you still want to benefit from automated node remediation.
+
+Considered allowing to target machineSets instead of using a label selector. This was discarded because of the reason above. Also there might be upper level controllers doing things that the MHC does not need to account for, e.g machineDeployment flipping machineSets for a rolling update. Therefore considering machines and label selectors as the fundamental operational entity results in a good and convenient level of flexibility and decoupling from other controllers.
+
+Considered a more strict short-circuiting mechanisim decoupled from the machine health checker i.e machine disruption budget analogous to pod disruption budget. This was discarded because it added an unjustified level of complexity and additional API resources. Instead we opt for a simpler approach and will consider more concrete feature requests that requires additional complexity based on real use feedback.
+
+## Upgrade Strategy
+This is an opt in feature supported by a new CRD and controller so no particular needs to handle upgrades for existing clusters. 
+
+## Additional Details
+
+### Test Plan [optional]
+Extensive unit testing for all the cases supported when remediating.
+
+e2e testing as part of the cluster-api e2e test suite.
+
+For failing early testing we could consider a test suite leveraging kubemark as a provider to simulate healthy/unhealthy nodes in a cloud agnostic manner without the need to bring up a real instance.
```

Comment:
@thebsdbox had built a fake provider for testing recently, we could potentially leverage that for this type of testing.